### ISSUE TITLE
Gamma variate gradients are incorrect

### ISSUE THREAD

**Issue by NeilGirdhar** [jax-ml/jax#27349]
### Description

```python
import jax.numpy as jnp
import jax.random as jr
from jax.test_util import check_grads
from tjax import JaxRealArray


def f(x: JaxRealArray) -> JaxRealArray:
    key = jr.key(0)
    samples = jr.gamma(key, x + 1.0, (1000,))
    return jnp.sum(samples, axis=0)


check_grads(f, (jnp.zeros(()),), order=1, atol=1e-4)
```
Should these be blocked (by adding a custom JVP that raises) until a suitable custom gradient can be provided?

### System info (python version, jaxlib version, accelerator, etc.)

Latest

**Comment by Gaurav23V** [comment]
Hi, I was able to reproduce the issue locally using the provided snippet.

```
import jax.numpy as jnp
import jax.random as jr
from jax import Array
from jax.test_util import check_grads

n = 10000

def f(x: Array) -> Array:
    key = jr.key(0)
    # Ensure x is float for gamma shape parameter 'a'
    a = jnp.asarray(x, dtype=float) + 1.0
    samples = jr.gamma(key, a, shape=(n, *a.shape))
    return jnp.sum(samples, axis=0) / n

print("Checking grads for gamma...")
# Using a small array to test broadcasting potentially
initial_value = jnp.array([0.5, 1.5])

check_grads(f, (initial_value,), order=1, modes=["rev"], atol=1e-4, rtol=1e-4)

print("check_grads call finished.") # Will likely not be reached on failure
```

The check grad function indeed fails the default tolerances.

I'd like to investigate this further. My initial thought is to examine the VJP rule defined for jax.random.gamma. It likely involves the digamma function (psi). I plan to:

1. Locate the VJP definition for jax.random.gamma in the source code.
2. Verify the mathematical correctness of the implemented gradient formula.
3. Investigate potential numerical stability issues in the calculation, especially concerning the digamma function or specific ranges of the input parameter a.

Would it be okay if I'm assigned this issue to work on a fix?

**Comment by NeilGirdhar** [comment]
If you decide to work on it, you may be interested in some personal communication I had with Michael Figurnov regarding his algorithm (see Implicit Reparameterization Gradients), which is used in Jax.

> * JAX implementation: https://github.com/jax-ml/jax/blob/540541a3d32fe6678aa1c208a4f5a9a697b92e2c/jax/_src/lax/special.py#L506-L524
> * XLA implementation which is what TensorFlow uses under the hood: https://github.com/openxla/xla/blob/5ed2ff5c07868d1a7486f4040f8b38936640268e/xla/hlo/builder/lib/math.cc#L1052-L1096
> 
> There is also a PyTorch implementation of [another method](https://arxiv.org/abs/1806.01851) to compute these derivatives. They use the same principle, but use a rational approximation for the derivative instead of computing it with forward-mode autodiff. At the time of writing the paper, my experiments showed it was less accurate and slower, although I suspect their method may be faster than autodiff on GPUs (I think I only tested with CPUs at the time).
> https://github.com/pytorch/pytorch/blob/a6ac6447b55bcf910dee5f925c2c17673f162a36/aten/src/ATen/native/Distributions.h#L305-L374

**Comment by Gaurav23V** [comment]
Thank you so much, @NeilGirdhar , for the  helpful pointers and links! I appreciate you sharing the context about the underlying algorithm and the specific implementations in JAX, XLA, and PyTorch.

I will carefully review the random_gamma_grad_impl function, the XLA/PyTorch code, and look into the "Implicit Reparameterization Gradients" method before proceeding further. I got a much clearer direction for investigating the check_grads failure. Thanks again!

**Comment by Gaurav23V** [comment]
Hi @NeilGirdhar  and maintainers,

Thank you again, Neil, for the helpful pointers regarding the implementation of the gamma gradient. Following up on my previous comment, I've investigated this further and wanted to share the findings.

1. Confirmed Discrepancy: I reproduced the check_grads failure and confirmed that the discrepancy between JAX's analytical gradient (VJP) and the numerical gradient persists even with a large number of samples (n=100000 in the test function), suggesting it's not solely due to sampling noise. The relative difference in the projected gradients compared by check_grads was around 1-1.5%.

2. Isolated random_gamma_grad_impl: I tested the internal function jax._src.lax.special.random_gamma_grad_impl directly for a specific input pair (a=1.5, x=1.1025...) that corresponds to the code path using _igamma_series (as indicated by tracing with n=5).

3. Significant Error Found: Comparing the output of random_gamma_grad_impl to a numerical approximation using scipy.special.gammainc (differentiated numerically w.r.t a) revealed a significant relative error of approximately 4.6% for this input.

4. Ruled out other factors:

- Checked jax.lax.digamma against scipy.special.digamma; they are very close (diff ~1.8e-7).

- Manually tightened the series convergence tolerance within the while_loop of _igamma_series. This did not change the final output, indicating the series was already converging based on its internal logic.

5. Hypothesis: Formula Error? Given that digamma is accurate and the series converges, things points towards a potential issue (either mathematical error or numerical instability) in the final formula used within _igamma_series when mode == IgammaMode.SAMPLE_DERIVATIVE: -(dans_da + ans * dlogax_da) * x / a (where ans and dans_da are the sums from the series, and dlogax_da = log(x) - digamma(a + 1)).

Could you perhaps confirm if this is the expected mathematical formula for this gradient component, or if there might be known numerical issues with this specific calculation within JAX? Verifying this formula seems like the necessary next step.

I'm still interested in helping to fix this if possible. Thanks!

**Comment by NeilGirdhar** [comment]
Sorry, I don't have time or the expertise to work on this, but I wish you the best of luck!!
